File: javascript/robot-name/robot-name.js
var names = [];

var Robot = function() {
  this.reset();
};

Robot.prototype.createName = function() {
  return 'xxyyy'.replace(/[xy]/g, function(c) {
    max = (c == 'x') ? 26 : 10;
    r = (Math.random() * max) | 0;
    return (c == 'x') ? String.fromCharCode(65 + r) : r.toString();
  });
}; 

Robot.prototype.createUniqueName = function() {
  var name;
  do {
    name = this.createName();
  } while(names[name] === true);
  names[name] = true;
  return name;
};
 
Robot.prototype.reset = function() {
  this.name = this.createUniqueName(); 
};

module.exports = Robot;

File: go/robot-name/robotname.go
package robotname

import (
	"math/rand"
	"sync"
	"time"
)

const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
const numbers = "0123456789"

// Robot represents a type of Robot.
type Robot struct {
	lk   sync.Mutex
	name string
}

type names struct {
	lk   sync.Mutex
	seen map[string]bool
}

type lockedRand struct {
	lk sync.Mutex
	r  *rand.Rand
}

func (lr *lockedRand) Intn(n int) (i int) {
	lr.lk.Lock()
	i = lr.r.Intn(n)
	lr.lk.Unlock()
	return
}

var robotNames = &names{seen: make(map[string]bool)}
var localRand *lockedRand

func init() {
	localRand = &lockedRand{r: rand.New(rand.NewSource(time.Now().UnixNano()))}
}

// Name returns the name of the Robot.
func (r *Robot) Name() (name string) {
	r.lk.Lock()
	if r.name == "" {
		r.name = robotNames.generateName()
	}
	name = r.name
	r.lk.Unlock()
	return
}

// Reset returns the Robot to factory settings.
func (r *Robot) Reset() {
	r.lk.Lock()
	r.name = ""
	r.lk.Unlock()
}

func (ns *names) generateName() (name string) {
	for {
		name = randString(2, letters) + randString(3, numbers)
		ns.lk.Lock()
		_, ok := ns.seen[name]
		if !ok {
			ns.seen[name] = true
		}
		ns.lk.Unlock()
		if !ok {
			break
		}
	}
	return
}

func randString(length int, charset string) string {
	bs := make([]byte, length)
	for i := range bs {
		bs[i] = charset[localRand.Intn(len(charset))]
	}
	return string(bs)
}
